## 1) Imports

In [2]:
import os
import numpy as np
import pandas as pd
import shutil

import tensorflow as tf
from xgboost import XGBClassifier

from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score, log_loss

from cleanlab.filter import find_label_issues
from collections import Counter

## 2) Training

In [2]:
labels_file = "image_labels.txt"
dataset_path = "dataset"

num_classes = 8
img_size = (224, 224)

batch_size = 32
n_splits = 5
random_state = 42

print("Config:")
print(" labels_file:", labels_file)
print(" dataset_path:", dataset_path)
print(" num_classes:", num_classes)
print(" img_size:", img_size)
print(" batch_size:", batch_size)
print(" n_splits:", n_splits)

pairs = []
with open(labels_file, "r", encoding="utf-8") as f:
    for line in f:
        parts = line.strip().split(": ")
        if len(parts) == 2:
            pairs.append((parts[0], int(parts[1])))

label_map = {fname: lbl for fname, lbl in pairs}

filenames = []
y_labels = []

for fname in os.listdir(dataset_path):
    if not fname.lower().endswith((".jpg", ".jpeg", ".png")):
        continue
    if fname not in label_map:
        continue
    filenames.append(fname)
    y_labels.append(label_map[fname])

filenames = np.array(filenames)
y_labels = np.array(y_labels, dtype=np.int64)

print("Samples:", len(filenames))
print("y_labels shape:", y_labels.shape)
print("label range:", (int(y_labels.min()), int(y_labels.max())))

if np.min(y_labels) < 0 or np.max(y_labels) >= num_classes:
    raise ValueError(
        f"Labels must be in [0, {num_classes-1}] but got range [{y_labels.min()}, {y_labels.max()}]."
    )

AUTOTUNE = tf.data.AUTOTUNE

def _load_and_preprocess(path):
    img_bytes = tf.io.read_file(path)
    img = tf.image.decode_image(img_bytes, channels=3, expand_animations=False)
    img = tf.image.resize(img, img_size, method="bilinear")
    img = tf.cast(img, tf.float32)
    img = tf.keras.applications.mobilenet_v2.preprocess_input(img)
    return img

full_paths = np.array([os.path.join(dataset_path, f) for f in filenames])

ds = tf.data.Dataset.from_tensor_slices(full_paths)
ds = ds.map(_load_and_preprocess, num_parallel_calls=AUTOTUNE)
ds = ds.batch(batch_size)
ds = ds.prefetch(AUTOTUNE)

print("\nDataset built.")

print("\nLoading MobileNetV2...")
base_model = tf.keras.applications.MobileNetV2(
    weights="imagenet",
    include_top=False,
    pooling="avg",
    input_shape=(img_size[0], img_size[1], 3),
)

print("Extracting embeddings (streaming batches)...")
X_features = base_model.predict(ds, verbose=1)

print("X_features shape:", X_features.shape)
print("Estimated RAM for embeddings (float32): ~", round(X_features.nbytes / (1024**2), 2), "MB")

print(f"🚀 Training XGBoost with {n_splits}-fold CV...")

pred_probs_cv = np.zeros((len(y_labels), num_classes), dtype=np.float32)

skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=random_state)
fold_acc, fold_ll = [], []

for fold, (train_idx, val_idx) in enumerate(skf.split(X_features, y_labels), start=1):
    clf = XGBClassifier(
        n_estimators=700,
        learning_rate=0.05,
        max_depth=6,
        subsample=0.8,
        colsample_bytree=0.8,
        reg_lambda=1.0,
        objective="multi:softprob",
        eval_metric="mlogloss",
        n_jobs=-1,
    )

    clf.fit(X_features[train_idx], y_labels[train_idx])
    probs = clf.predict_proba(X_features[val_idx])

    if probs.shape[1] != num_classes:
        raise RuntimeError(f"Expected probs with {num_classes} classes, got {probs.shape[1]}")

    pred_probs_cv[val_idx] = probs

    y_val = y_labels[val_idx]
    y_pred = np.argmax(probs, axis=1)

    acc = accuracy_score(y_val, y_pred)
    ll = log_loss(y_val, probs, labels=list(range(num_classes)))

    fold_acc.append(acc)
    fold_ll.append(ll)
    print(f"  Fold {fold}: acc={acc:.4f}, logloss={ll:.4f}")

print(f"CV mean acc={np.mean(fold_acc):.4f} ± {np.std(fold_acc):.4f}")
print(f"CV mean logloss={np.mean(fold_ll):.4f} ± {np.std(fold_ll):.4f}")

row_sums = pred_probs_cv.sum(axis=1)
print("prob row sum stats:", float(row_sums.min()), float(row_sums.mean()), float(row_sums.max()))

Config:
 labels_file: image_labels.txt
 dataset_path: dataset
 num_classes: 8
 img_size: (224, 224)
 batch_size: 32
 n_splits: 5
Samples: 8707
y_labels shape: (8707,)
label range: (0, 7)

Dataset built.

Loading MobileNetV2...
9406464/9406464 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
Extracting embeddings (streaming batches)...
273/273 ━━━━━━━━━━━━━━━━━━━━ 71s 256ms/step 
X_features shape: (8707, 1280)
Estimated RAM for embeddings (float32): ~ 42.51 MB
🚀 Training XGBoost with 5-fold CV...
  Fold 1: acc=0.5677, logloss=1.1791
  Fold 2: acc=0.5947, logloss=1.1831
  Fold 3: acc=0.5669, logloss=1.2648
  Fold 4: acc=0.5916, logloss=1.1812
  Fold 5: acc=0.5893, logloss=1.1795
CV mean acc=0.5821 ± 0.0122
CV mean logloss=1.1975 ± 0.0337
prob row sum stats: 0.9999998807907104 1.0 1.0000001192092896


Due to time constraints for training, I decided to use MobileNet for feature extraction and XGBoost for classification, as both are very fast and achieve the best performance. Other models did not perform as well; for example, some models I tried to build entirely from scratch had poor accuracy, causing many good images to be incorrectly identified as image errors. Some other models were prone to overfitting and failed to detect any image errors. To improve model robustness, I applied K-fold cross-validation and, due to training time limitations, set K = 5.

## 3) Cleanlab: find suspected label issues

In [3]:
print("Running Cleanlab...")

issues = find_label_issues(
    labels=y_labels,
    pred_probs=pred_probs_cv,
    return_indices_ranked_by="self_confidence",
)

print(f"Total suspected label issues: {len(issues)}")

Running Cleanlab...
Total suspected label issues: 2378


Applying cleanlab to identify the image errors

## 4) Review top issues + export CSV

In [4]:
top_k = min(50, len(issues))

if top_k == 0:
    print("No label issues found by Cleanlab.")
    print("If you still suspect label noise, try:")
    print("- n_splits=10")
    print("- smaller/regularized XGBoost")
    print("- calibrating probabilities (e.g., temperature scaling) before Cleanlab")
else:
    rows = []
    print(f"Top {top_k} most suspicious examples:")

    for rank, idx in enumerate(issues[:top_k], start=1):
        given = int(y_labels[idx])
        probs = pred_probs_cv[idx]
        pred = int(np.argmax(probs))

        row = {
            "rank": rank,
            "index": int(idx),
            "filename": filenames[idx],
            "given_label": given,
            "pred_label": pred,
            "p_given": float(probs[given]),
            "p_pred": float(probs[pred]),
        }
        rows.append(row)

        print(
            f"{rank:02d}. idx={row['index']} file={row['filename']} "
            f"given={row['given_label']} pred={row['pred_label']} "
            f"p(given)={row['p_given']:.4f} p(pred)={row['p_pred']:.4f}"
        )

    df = pd.DataFrame(rows)
    df.to_csv("suspected_label_issues.csv", index=False)
    print("Saved: suspected_label_issues.csv")

Top 50 most suspicious examples:
01. idx=7203 file=Stage_7_u2356551_10_.JPG given=6 pred=5 p(given)=0.0002 p(pred)=0.9940
02. idx=7196 file=Stage_7_u2356551_03_.JPG given=6 pred=5 p(given)=0.0002 p(pred)=0.9896
03. idx=4184 file=Stage_4_U2355526 (4).jpg given=3 pred=4 p(given)=0.0003 p(pred)=0.9969
04. idx=742 file=Stage_1_2359270_8.jpeg given=0 pred=1 p(given)=0.0004 p(pred)=0.9817
05. idx=7202 file=Stage_7_u2356551_09_.JPG given=6 pred=5 p(given)=0.0004 p(pred)=0.9909
06. idx=7195 file=Stage_7_u2356551_02_.JPG given=6 pred=5 p(given)=0.0006 p(pred)=0.9911
07. idx=6879 file=Stage_7_2384998_4.JPG given=6 pred=7 p(given)=0.0006 p(pred)=0.9857
08. idx=4190 file=Stage_4_U2355526.jpg given=3 pred=4 p(given)=0.0006 p(pred)=0.9954
09. idx=4188 file=Stage_4_U2355526 (8).jpg given=3 pred=4 p(given)=0.0006 p(pred)=0.9950
10. idx=8237 file=stage_8_U2357521(9).jpg given=7 pred=2 p(given)=0.0007 p(pred)=0.7069
11. idx=2953 file=Stage_3_2481197_10.jpg given=2 pred=3 p(given)=0.0009 p(pred)=0.9882
1

## Label errors:
**Explain what kind of errors you found in the dataset.**
1. Image not related to the topic.
2. Incorrect hand pose.
4. AI-generated image.
5. Image contains many distracting objects.
6. Includes full body instead of only the hand.
7. White regions due to resizing
8. Only part of the hand appears in the image, providing insufficient visual features for reliable identification (only the palm is visible).
9. The image of Stage 1 shows hand washing without water.
10. The hand is holding the soap dispenser.

## 5) train the same model on the new dataset (after remove all image errors)

In [5]:
ACTION = "move"
quarantine_subdir = "_cleanlab_label_issues"
top_k_to_remove = None 

if "issues" not in globals():
    raise NameError("Variable `issues` not found. Run the Cleanlab step first to compute `issues`.")
if "filenames" not in globals():
    raise NameError("Variable `filenames` not found. Run the data loading steps first.")
if "dataset_path" not in globals():
    raise NameError("Variable `dataset_path` not found. Run the config step first.")
if "labels_file" not in globals():
    raise NameError("Variable `labels_file` not found. Run the config step first.")

if len(issues) == 0:
    print("No issues to remove.")
else:
    quarantine_dir = os.path.join(dataset_path, quarantine_subdir)

    idxs = issues if top_k_to_remove is None else issues[:top_k_to_remove]
    files_to_handle = [str(filenames[i]) for i in idxs]

    print(f"Will {ACTION} {len(files_to_handle)} files.")

    if ACTION == "move":
        os.makedirs(quarantine_dir, exist_ok=True)

    handled = 0
    missing = 0

    for fname in files_to_handle:
        src = os.path.join(dataset_path, fname)
        if not os.path.exists(src):
            missing += 1
            continue

        if ACTION == "move":
            dst = os.path.join(quarantine_dir, fname)
            if os.path.exists(dst):
                base, ext = os.path.splitext(fname)
                k = 1
                while True:
                    dst2 = os.path.join(quarantine_dir, f"{base}__{k}{ext}")
                    if not os.path.exists(dst2):
                        dst = dst2
                        break
                    k += 1
            shutil.move(src, dst)

        elif ACTION == "delete":
            os.remove(src)

        else:
            raise ValueError("ACTION must be 'move' or 'delete'")

        handled += 1

    print(f"✔ Done. Handled: {handled}, Missing: {missing}")
    if ACTION == "move":
        print("Moved files to:", quarantine_dir)

    cleaned_labels_file = "image_labels.cleaned.txt"
    remove_set = set(files_to_handle)
    kept_lines = []

    with open(labels_file, "r", encoding="utf-8") as f:
        for line in f:
            parts = line.strip().split(": ")
            if len(parts) != 2:
                continue
            fname = parts[0]
            if fname in remove_set:
                continue
            kept_lines.append(line.strip())

    with open(cleaned_labels_file, "w", encoding="utf-8") as f:
        f.write("\n".join(kept_lines) + ("\n" if kept_lines else ""))

    print("Wrote cleaned labels file:", cleaned_labels_file)
    print("Next run: set labels_file = 'image_labels.cleaned.txt'")

Will move 2378 files.
✔ Done. Handled: 2378, Missing: 0
Moved files to: dataset\_cleanlab_label_issues
Wrote cleaned labels file: image_labels.cleaned.txt
Next run: set labels_file = 'image_labels.cleaned.txt'


Try to remove all image errors, because it performs best when all image errors are removed.

In [6]:
labels_file = "image_labels.cleaned.txt"
dataset_path = "dataset"

num_classes = 8
img_size = (224, 224)

batch_size = 32
n_splits = 5
random_state = 42

print("Config:")
print(" labels_file:", labels_file)
print(" dataset_path:", dataset_path)
print(" num_classes:", num_classes)
print(" img_size:", img_size)
print(" batch_size:", batch_size)
print(" n_splits:", n_splits)

pairs = []
with open(labels_file, "r", encoding="utf-8") as f:
    for line in f:
        parts = line.strip().split(": ")
        if len(parts) == 2:
            pairs.append((parts[0], int(parts[1])))

label_map = {fname: lbl for fname, lbl in pairs}

filenames = []
y_labels = []

for fname in os.listdir(dataset_path):
    if not fname.lower().endswith((".jpg", ".jpeg", ".png")):
        continue
    if fname not in label_map:
        continue
    filenames.append(fname)
    y_labels.append(label_map[fname])

filenames = np.array(filenames)
y_labels = np.array(y_labels, dtype=np.int64)

print("Samples:", len(filenames))
print("y_labels shape:", y_labels.shape)
print("label range:", (int(y_labels.min()), int(y_labels.max())))

if np.min(y_labels) < 0 or np.max(y_labels) >= num_classes:
    raise ValueError(
        f"Labels must be in [0, {num_classes-1}] but got range [{y_labels.min()}, {y_labels.max()}]."
    )

AUTOTUNE = tf.data.AUTOTUNE

def _load_and_preprocess(path):
    img_bytes = tf.io.read_file(path)
    img = tf.image.decode_image(img_bytes, channels=3, expand_animations=False)
    img = tf.image.resize(img, img_size, method="bilinear")
    img = tf.cast(img, tf.float32)
    img = tf.keras.applications.mobilenet_v2.preprocess_input(img)
    return img

full_paths = np.array([os.path.join(dataset_path, f) for f in filenames])

ds = tf.data.Dataset.from_tensor_slices(full_paths)
ds = ds.map(_load_and_preprocess, num_parallel_calls=AUTOTUNE)
ds = ds.batch(batch_size)
ds = ds.prefetch(AUTOTUNE)

print("\nDataset built.")

print("\nLoading MobileNetV2...")
base_model = tf.keras.applications.MobileNetV2(
    weights="imagenet",
    include_top=False,
    pooling="avg",
    input_shape=(img_size[0], img_size[1], 3),
)

print("Extracting embeddings (streaming batches)...")
X_features = base_model.predict(ds, verbose=1)

print("X_features shape:", X_features.shape)
print("Estimated RAM for embeddings (float32): ~", round(X_features.nbytes / (1024**2), 2), "MB")

print(f"🚀 Training XGBoost with {n_splits}-fold CV...")

pred_probs_cv = np.zeros((len(y_labels), num_classes), dtype=np.float32)

skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=random_state)
fold_acc, fold_ll = [], []

for fold, (train_idx, val_idx) in enumerate(skf.split(X_features, y_labels), start=1):
    clf = XGBClassifier(
        n_estimators=700,
        learning_rate=0.05,
        max_depth=6,
        subsample=0.8,
        colsample_bytree=0.8,
        reg_lambda=1.0,
        objective="multi:softprob",
        eval_metric="mlogloss",
        n_jobs=-1,
    )

    clf.fit(X_features[train_idx], y_labels[train_idx])
    probs = clf.predict_proba(X_features[val_idx])

    if probs.shape[1] != num_classes:
        raise RuntimeError(f"Expected probs with {num_classes} classes, got {probs.shape[1]}")

    pred_probs_cv[val_idx] = probs

    y_val = y_labels[val_idx]
    y_pred = np.argmax(probs, axis=1)

    acc = accuracy_score(y_val, y_pred)
    ll = log_loss(y_val, probs, labels=list(range(num_classes)))

    fold_acc.append(acc)
    fold_ll.append(ll)
    print(f"  Fold {fold}: acc={acc:.4f}, logloss={ll:.4f}")

print(f"CV mean acc={np.mean(fold_acc):.4f} ± {np.std(fold_acc):.4f}")
print(f"CV mean logloss={np.mean(fold_ll):.4f} ± {np.std(fold_ll):.4f}")

row_sums = pred_probs_cv.sum(axis=1)
print("prob row sum stats:", float(row_sums.min()), float(row_sums.mean()), float(row_sums.max()))

Config:
 labels_file: image_labels.cleaned.txt
 dataset_path: dataset
 num_classes: 8
 img_size: (224, 224)
 batch_size: 32
 n_splits: 5
Samples: 6329
y_labels shape: (6329,)
label range: (0, 7)

Dataset built.

Loading MobileNetV2...
Extracting embeddings (streaming batches)...
198/198 ━━━━━━━━━━━━━━━━━━━━ 47s 233ms/step
X_features shape: (6329, 1280)
Estimated RAM for embeddings (float32): ~ 30.9 MB
🚀 Training XGBoost with 5-fold CV...
  Fold 1: acc=0.7077, logloss=0.8134
  Fold 2: acc=0.7172, logloss=0.8037
  Fold 3: acc=0.7322, logloss=0.7685
  Fold 4: acc=0.7551, logloss=0.7313
  Fold 5: acc=0.7209, logloss=0.7919
CV mean acc=0.7267 ± 0.0163
CV mean logloss=0.7818 ± 0.0293
prob row sum stats: 0.9999998807907104 1.0 1.0000001192092896


After removing all image errors, the accuracy increased from 0.58 to 0.78, which is a significant improvement.

In [3]:

label_file = "image_labels.cleaned.txt"


stage_counter = Counter()


with open(label_file, "r", encoding="utf-8") as f:
    for line in f:
        line = line.strip()
        if not line:
            continue

        parts = line.split(":")
        if len(parts) != 2:
            continue
        filename, label_str = parts
        try:
            label = int(label_str.strip())
            stage_counter[label] += 1
        except:
            continue


for stage in range(8):
    print(f"Stage {stage+1}: {stage_counter.get(stage, 0)} images")


Stage 1: 947 images
Stage 2: 694 images
Stage 3: 723 images
Stage 4: 748 images
Stage 5: 787 images
Stage 6: 727 images
Stage 7: 795 images
Stage 8: 908 images


**List the total number of images left in each class/stage after the label error handling**

<br>

<ol>
  <li>Stage 1: 947 images</li>
  <li>Stage 2: 694 images</li>
  <li>Stage 3: 723 images</li>
  <li>Stage 4: 748 images</li>
  <li>Stage 5: 787 images</li>
  <li>Stage 6: 727 images</li>
  <li>Stage 7: 795 images</li>
  <li>Stage 8: 908 images</li>
</ol>